### PR TITLE
[Messenger] make Envelope first class citizen for middleware handlers

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR sits on top of #28909 so that only the 2nd commit should be reviewed for now, until I rebase it.
This idea has already been proposed and rejected in #27322.
Yet, now that I've reviewed in depth the component, I politely but strongly suggest to reconsider.
Middleware handlers are the central piece of the routing layer.
When `dispatch()` accepts `object|Envelope`, it's because it sits on the outer boundary of the component. `handle()` on the contrary plugs inside its core routing logic, so that it's very legitimate to require them to deal with `Envelope` objects. Yes, some middleware care only about the message inside. That's fine calling `getMessage()` for them.

Right now, we built a complex magic layer that acts as a band-aid *just* to save this call to `getMessage()`. For middleware that want the envelope, we require an additional interface that magically *changes* the expected argument. That's very fragile design: it breaks the `L` in `SOLID`...

Looking at the diff stat, this is most natural.

### CODE REVIEW

**Review by ogizanagi — APPROVED** [review]
Should we remove `Envelope::wrap()` as in original PR or do you consider it still relevant? We can decide this in another PR though.

**Review by weaverryan — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
> Should we remove Envelope::wrap()

yes! same for `Envelope->withMessage()` as its unused anymore. Done.

**Comment by nicolas-grekas** [comment]
(rebased, ready)

**Review by sroze — APPROVED** [review]
You've made @ogizanagi happy. 😄

**Comment by sroze** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/Messenger/Middleware/MiddlewareInterface.php [hunk 1]
```diff
@@ -11,13 +11,23 @@
...
+     * @param callable|NextInterface $next
      */
-    public function handle($message, callable $next): void;
+    public function handle(Envelope $envelope, callable $next): void;
+}
+
+/**
+ * @internal
+ */
+interface NextInterface
```
Comment (Koc) [inline src/Symfony/Component/Messenger/Middleware/MiddlewareInterface.php]: Where it used?

**Pull request merged**